OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs

```diff
@@ -0,0 +1,312 @@
+                               "\"Exception while fetching data (/customer) : " +
+                               "java.lang.RuntimeException: could not resolve _fs_user_id = invalid-user\"," +
+                               "\"locations\":[{\"line\":1,\"column\":8}],\"path\":[\"customer\"]," +
+                               "\"extensions\":{\"classification\":\"DataFetchingException\"}}]," +
+                               "\"data\":{\"customer\":null}}";
+            _mockOdpClient.Setup(
+                    c => c.QuerySegments(It.IsAny<QuerySegmentsParameters>())).
+                Returns(responseData);
+            var manager = new GraphQLManager(_mockLogger.Object);
+
+            var segments = manager.FetchSegments(
+                VALID_ODP_PUBLIC_KEY,
+                ODP_GRAPHQL_URL,
+                FS_USER_ID,
+                "invalid-user", // invalid user
+                _segmentsToCheck);
+
+            Assert.IsTrue(segments.Length == 0);
+            _mockLogger.Verify(l => l.Log(LogLevel.WARN, It.IsAny<string>()), Times.Once);
+        }
+
+        [Test]
+        public void ShouldHandleOtherException()
+        {
+            var responseData = "{\"errors\":[{\"message\":\"Validation error of type " +
+                               "UnknownArgument: Unknown field argument not_real_userKey @ " +
+                               "'customer'\",\"locations\":[{\"line\":1,\"column\":17}]," +
+                               "\"extensions\":{\"classification\":\"ValidationError\"}}]}";
+
+            _mockOdpClient.Setup(
+                    c => c.QuerySegments(It.IsAny<QuerySegmentsParameters>())).
+                Returns(responseData);
+            var manager = new GraphQLManager(_mockLogger.Object, _mockOdpClient.Object);
+
+            var segments = manager.FetchSegments(
+                VALID_ODP_PUBLIC_KEY,
+                ODP_GRAPHQL_URL,
+                "not_real_userKey",
+                "tester-101",
+                _segmentsToCheck);
+
+            Assert.IsTrue(segments.Length == 0);
+            _mockLogger.Verify(l => l.Log(LogLevel.WARN, It.IsAny<string>()), Times.Once);
+        }
+
+        [Test]
+        public void ShouldHandleBadResponse()
+        {
+            var responseData = "{\"data\":{ }}";
+            _mockOdpClient.Setup(
+                    c => c.QuerySegments(It.IsAny<QuerySegmentsParameters>())).
+                Returns(responseData);
+            var manager = new GraphQLManager(_mockLogger.Object, _mockOdpClient.Object);
+
+            var segments = manager.FetchSegments(
+                VALID_ODP_PUBLIC_KEY,
+                ODP_GRAPHQL_URL,
+                "not_real_userKey",
+                "tester-101",
+                _segmentsToCheck);
+
+            Assert.IsNull(segments);
+            _mockLogger.Verify(l => l.Log(LogLevel.WARN, It.IsAny<string>()), Times.Never);
+        }
+
+        [Test]
+        public void ShouldHandleNetworkError() { }
```

Comment:
Any way to inject a network error?

Reply:
I did forget to develop this test. Apologies.

I looked Friday and again this morning. I can't seem to find a way to cause or simulate a network error within a test other than to inject a desired response status code.  As an alternative, I've restructured to handle 4XX and 5XX responses.